lib/Dialect/FIRRTL/Transforms/LowerAnnotations.cpp

```diff
@@ -0,0 +1,661 @@
+    }
+    if (op->getNumResults() == 0)
+      return FIRRTLType();
+    return op->getResult(0).getType().cast<FIRRTLType>().getSubTypeByFieldID(
+        fieldIdx);
+  }
+};
+
+struct AnnoPathValue {
+  SmallVector<InstanceOp> instances;
+  BaseUnion ref;
+
+  AnnoPathValue() = default;
+  AnnoPathValue(CircuitOp op) : ref(op) {}
+  AnnoPathValue(Operation *op) : ref(op) {}
+  AnnoPathValue(const SmallVectorImpl<InstanceOp> &insts, BaseUnion b)
+      : instances(insts.begin(), insts.end()), ref(b) {}
+
+  bool isLocal() const { return instances.empty(); }
+  template <typename T>
+  bool isOpOfType() const {
+    if (!ref || ref.isPort())
+      return false;
+    return isa<T>(ref.op);
+  }
+};
+
+} // namespace
+
+static bool hasName(StringRef name, Operation *op) {
+  return TypeSwitch<Operation *, bool>(op)
+      .Case<InstanceOp, MemOp, NodeOp, RegOp, RegResetOp, WireOp, CMemOp,
+            SMemOp, MemoryPortOp>([&](auto nop) {
+        if (nop.name() == name)
+          return true;
+        return false;
+      })
+      .Default([](auto &) { return false; });
+}
+
+static BaseUnion findNamedThing(StringRef name, Operation *op) {
+  BaseUnion retval;
+  auto nameChecker = [name, &retval](Operation *op) -> WalkResult {
+    if (isa<FModuleOp, FExtModuleOp>(op)) {
+      // Check the ports.
+      auto ports = getModulePortInfo(op);
+      for (size_t i = 0, e = ports.size(); i != e; ++i)
+        if (ports[i].name.getValue() == name) {
+          retval = BaseUnion{op, i};
+          return WalkResult::interrupt();
+        }
+      return WalkResult::advance();
+    }
+    if (hasName(name, op)) {
+      retval = BaseUnion{op};
+      return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  };
+  op->walk(nameChecker);
+  return retval;
+}
+
+static void addNamedAttr(Operation *op, StringRef name, StringAttr value) {
+  op->setAttr(name, value);
+}
+
+static void addNamedAttr(Operation *op, StringRef name) {
+  op->setAttr(name, BoolAttr::get(op->getContext(), true));
+}
+
+static ArrayAttr getAnnotationsFrom(Operation *op) {
+  if (auto annots = op->getAttrOfType<ArrayAttr>(getAnnotationAttrName()))
+    return annots;
+  return ArrayAttr::get(op->getContext(), {});
+}
+
+static void addAnnotation(Operation *op, Annotation anno) {
+  SmallVector<Attribute> newAnnos;
+  for (auto old : getAnnotationsFrom(op))
+    newAnnos.push_back(old);
+  newAnnos.push_back(anno.getDict());
+  op->setAttr(getAnnotationAttrName(),
+              ArrayAttr::get(op->getContext(), newAnnos));
+}
+
+// Returns remainder of path if circuit is the correct circuit
+LogicalResult parseAndCheckCircuit(StringRef &path, CircuitOp circuit) {
+  if (path.startswith("~"))
+    path = path.drop_front();
+
+  // Any non-trivial target must start with a circuit name.
+  StringRef name;
+  std::tie(name, path) = path.split('|');
+  // ~ or ~Foo
+  if (name.empty() || name == circuit.name())
+    return success();
```

Comment:
```suggestion
    return success();
    // foo.foo
  std::tie(name, path) = path.split('.');
  if (name.empty() || name == circuit.name())
    return success();  
```
To handle the case: 
``
circuit Foo: %[[{"class":"circt.test","a":"a","target":"Foo.Foo"}]]
``

Reply:
We need to handle the arbitrarily deep dot form, which means we need to run the target canonicalization Schuyler had before.